corehq/util/models.py

```diff
@@ -36,10 +36,52 @@ class BounceSubType(object):
     )
 
 
+# The number of General bounces we accept before rejecting an email hard
+GENERAL_BOUNCE_THRESHOLD = 3
+
+
 class BouncedEmail(models.Model):
     created = models.DateTimeField(auto_now_add=True)
     email = models.EmailField(db_index=True, unique=True)
 
+    @staticmethod
+    def get_hard_bounced_emails(list_of_emails):
+        # these are any Bounced Email Records we have
+        bounced_emails = set(
+            BouncedEmail.objects.filter(email__in=list_of_emails).values_list(
+                'email', flat=True)
+        )
+
+        # These are emails that were marked as Suppressed or Undetermined
+        # by SNS metadata, meaining they definitely hard bounced
+        bad_emails = set(
+            PermanentBounceMeta.objects.filter(sub_type__in=[
+                BounceSubType.UNDETERMINED, BounceSubType.SUPPRESSED
+            ], bounced_email__email__in=bounced_emails).values_list(
+                'bounced_email__email', flat=True)
+        )
+
+        # These are definite complaints against us
+        complaints = set(
+            ComplaintBounceMeta.objects.filter(
+                bounced_email__email__in=bounced_emails.difference(bad_emails)
+            ).values_list('bounced_email__email', flat=True)
+        )
+        bad_emails.union(complaints)
```

Comment:
I think you meant this to be an update operation, but `.union` just returns a new set without mutating the original. You might try one of following (which I think are equivalent to each other):

```
bad_emails |= complaints
```
or
```
bad_emails.update(complaints)
```

Reply:
so I thought that as well...but i think it actually works as intended otherwise the tests would be failing no? 👀 I will use `update` regardless just so it's 100% clear.

Reply:
Yeah I'm not sure what's up with the tests, but this line for sure wasn't updating bad_emails the way it was:

```
In [1]: x = {1, 2, 3}

In [2]: x.union({4, 5, 6})
Out[2]: {1, 2, 3, 4, 5, 6}

In [3]: x
Out[3]: {1, 2, 3}

In [4]: x.update({4, 5, 6})

In [5]: x
Out[5]: {1, 2, 3, 4, 5, 6}
```